libr/bin/p/bin_dyldcache.c

```diff
@@ -794,6 +794,68 @@ static int string_contains(const void *a, const void *b) {
 	return !strstr ((const char*) a, (const char*) b);
 }
 
+static HtPP * create_path_to_index(RBuffer *cache_buf, cache_img_t *img, cache_hdr_t *hdr) {
+	HtPP *path_to_idx = ht_pp_new0 ();
+	if (!path_to_idx) {
+		return NULL;
+	}
+	ut64 i;
+	for (i = 0; i != hdr->imagesCount; i++) {
+		char file[256];
+		if (r_buf_read_at (cache_buf, img[i].pathFileOffset, (ut8*) &file, sizeof (file)) != sizeof (file)) {
+			continue;
+		}
+		file[255] = 0;
+		const char *key = sdb_fmt ("%s", file);
+		ht_pp_insert (path_to_idx, key, (void*) i);
+	}
+
+	return path_to_idx;
+}
+
+static void carve_deps_at_address(RBuffer *cache_buf, cache_img_t *img, cache_hdr_t *hdr, cache_map_t *maps, HtPP *path_to_idx, ut64 address, int *deps) {
+	ut64 pa = va2pa (address, hdr, maps, cache_buf, 0, NULL, NULL);
+	if (pa == UT64_MAX) {
+		return;
+	}
+	struct MACH0_(mach_header) mh;
+	if (r_buf_fread_at (cache_buf, pa, (ut8*) &mh, "8i", 1) != sizeof (struct MACH0_(mach_header))) {
+		return;
+	}
+	if (mh.magic != MH_MAGIC_64 || mh.sizeofcmds == 0) {
+		return;
+	}
+	ut64 cmds_at = pa + sizeof (struct MACH0_(mach_header));
+	ut8 *cmds = malloc (mh.sizeofcmds);
+	if (!cmds || r_buf_read_at (cache_buf, cmds_at, cmds, mh.sizeofcmds) != mh.sizeofcmds) {
+		goto beach;
+	}
+	ut8 *cursor = cmds;
+	ut8 *end = cmds + mh.sizeofcmds;
+	while (cursor < end) {
+		ut32 cmd = r_read_le32 (cursor);
+		ut32 cmdsize = r_read_le32 (cursor + sizeof (ut32));
+		if (cmd == LC_LOAD_DYLIB ||
+				cmd == LC_LOAD_WEAK_DYLIB ||
+				cmd == LC_REEXPORT_DYLIB ||
+				cmd == LC_LOAD_UPWARD_DYLIB) {
+			bool found;
+			const char *key = sdb_fmt ("%s", cursor + 24);
```

Comment:
no need to fmt this. just key = cursor + 24

Reply:
also, check if cursor +24 < end and probably set the last byte of cmds to null to avoid OOB reads